lib/srv/app/aws/handler.go

```diff
@@ -17,211 +17,171 @@ limitations under the License.
 	if err != nil {
-		s.Log.WithError(err).Warn("Failed to emit audit event.")
+		return trace.Wrap(err)
 	}
-	return resp, nil
-}
 
-func (s *SigningService) formatForwardResponseError(rw http.ResponseWriter, r *http.Request, err error) {
-	common.SetTeleportAPIErrorHeader(rw, err)
-
-	switch trace.Unwrap(err).(type) {
-	case *trace.BadParameterError:
-		s.Log.Debugf("Failed to process request: %v.", err)
-		rw.WriteHeader(http.StatusBadRequest)
-	case *trace.AccessDeniedError:
-		s.Log.Infof("Failed to process request: %v.", err)
-		rw.WriteHeader(http.StatusForbidden)
-	default:
-		s.Log.Warnf("Failed to process request: %v.", err)
-		rw.WriteHeader(http.StatusInternalServerError)
+	// rewrite headers before signing the request to avoid signature validation problems.
+	unsignedReq, err := rewriteRequest(req, re)
+	if err != nil {
+		return trace.Wrap(err)
 	}
-}
 
-// prepareSignedRequest creates a new HTTP request and rewrites the header from the original request and returns a new
-// HTTP request signed by STS AWS API.
-func (s *SigningService) prepareSignedRequest(r *http.Request, payload []byte, re *endpoints.ResolvedEndpoint, sessionCtx *common.SessionContext) (*http.Request, error) {
-	url, err := urlForResolvedEndpoint(r, re)
+	signedReq, err := s.SignRequest(unsignedReq,
+		&awsutils.SigningCtx{
+			SigningName:   re.SigningName,
+			SigningRegion: re.SigningRegion,
+			Expiry:        sessCtx.Identity.Expires,
+			SessionName:   sessCtx.Identity.Username,
+			AWSRoleArn:    sessCtx.Identity.RouteToApp.AWSRoleARN,
+			AWSExternalID: sessCtx.App.GetAWSExternalID(),
+		})
 	if err != nil {
-		return nil, trace.Wrap(err)
+		return trace.Wrap(err)
+	}
+	recorder := httplib.NewResponseStatusRecorder(w)
+	s.fwd.ServeHTTP(recorder, signedReq)
+
+	var auditErr error
+	if isDynamoDBEndpoint(re) {
+		auditErr = sessCtx.Audit.OnDynamoDBRequest(unsignedReq.Context(), sessCtx, unsignedReq, recorder.Status(), re)
+	} else {
+		auditErr = sessCtx.Audit.OnRequest(unsignedReq.Context(), sessCtx, unsignedReq, recorder.Status(), re)
 	}
-	reqCopy, err := http.NewRequest(r.Method, url, bytes.NewReader(payload))
 	if err != nil {
-		return nil, trace.Wrap(err)
+		// log but don't return the error, because we already handed off request/response handling to the oxy forwarder.
+		s.Log.WithError(auditErr).Warn("Failed to emit audit event.")
 	}
-	rewriteHeaders(r, reqCopy)
-	// Sign the copy of the request.
-	signer := awsutils.NewSigner(s.getSigningCredentials(s.Session, sessionCtx), re.SigningName)
-	_, err = signer.Sign(reqCopy, bytes.NewReader(payload), re.SigningName, re.SigningRegion, s.Clock.Now())
+	return nil
+}
+
+// rewriteRequest rewrites a request to remove Teleport reserved headers, sets the url, and sets host.
+func rewriteRequest(r *http.Request, re *endpoints.ResolvedEndpoint) (*http.Request, error) {
+	// shallow copy request and make a deep copy for header modification.
+	outReq := &http.Request{}
+	*outReq = *r
```

Comment:
can't you just use `r.Clone()`?
I am afraid that some other fields are missing from this clone and can cause bugs in the future.

It also clones the `r.URL` so you can safely manipulate it in `urlForResolvedEndpoint`

Reply:
good catch, it "works" but definitely looks bug-prone.

How's this instead:
```golang
// rewriteRequest rewrites a request to remove Teleport reserved headers, sets the url, and sets host.
func rewriteRequest(ctx context.Context, r *http.Request, re *endpoints.ResolvedEndpoint) (*http.Request, error) {
	u, err := urlForResolvedEndpoint(r, re)
	if err != nil {
		return nil, trace.Wrap(err)
	}
	outReq, err := http.NewRequestWithContext(ctx, r.Method, u.String(), r.Body)
	if err != nil {
		return nil, trace.Wrap(err)
	}
	outReq.Header = r.Header.Clone()
...
```

Reply:
`ctx` in this case would be the app server close context by the way.

Reply:
See above comments. I will change this to use clone instead, as you suggested. This stuff is subtle 😑

Reply:
updated to clone and assign the clone's body to a new readcloser.